polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp

```diff
@@ -75,93 +76,264 @@ void Uniformity::print(raw_ostream &os) const { os << *this; }
+    }
+
+    // If we can't yet compute the mods/pMods operands uniformity, bail out.
+    if (!canComputeUniformity(*mods)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2)
+                 << "Reaching def operand(s) uniformity not yet initialized\n");
       return;
     }
 
-    TypeSwitch<MemoryEffects::Effect *>(effect.getEffect())
-        .Case<MemoryEffects::Read>([](auto) {
-          // A read operation yields uniform result(s) iff the (potentially)
-          // reaching definitions of all its operands are uniform.
-          assert(false && "TODO");
+    // If any modifiers or potential modifiers of the value loaded store a value
+    // that is unknown/non-uniform the result(s) of the load are also
+    // unknown/non-uniform.
+    if (anyModifierUniformityIs(*mods, Uniformity::Kind::Unknown)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2)
+                 << "Reaching def has unknown uniformity\n");
+      return propagateAllIfChanged(op->getResults(), Uniformity::getUnknown());
+    }
+    if (anyModifierUniformityIs(*mods, Uniformity::Kind::NonUniform)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2) << "Reaching def is non-uniform\n");
+      return propagateAllIfChanged(op->getResults(),
+                                   Uniformity::getNonUniform());
+    }
+  }
+
+  LLVM_DEBUG(llvm::dbgs().indent(2) << "Memory effects analyzed\n");
+  return propagateAllIfChanged(op->getResults(), Uniformity::getUniform());
+}
+
+SmallVector<Value> UniformityAnalysis::collectBranchConditions(
+    const ReachingDefinition::ModifiersTy &mods) {
+  SmallVector<Value> conditions;
+  for (const Definition &mod : mods) {
+    if (!mod.isOperation())
+      continue;
+
+    SetVector<RegionBranchOpInterface> enclosingBranches =
+        getParentsOfType<RegionBranchOpInterface>(
+            *mod.getOperation()->getBlock());
+    for (RegionBranchOpInterface branchOp : enclosingBranches) {
+      LLVM_DEBUG(llvm::dbgs().indent(2) << "branchOp: " << branchOp << "\n");
+      conditions.push_back(getCondition(branchOp));
+    }
+  }
+  return conditions;
+}
+
+bool UniformityAnalysis::canComputeUniformity(
+    const ReachingDefinition::ModifiersTy &mods) {
+  return llvm::all_of(mods, [&](const Definition &def) {
+    if (!def.isOperation())
+      return true;
+    return TypeSwitch<Operation *, bool>(def.getOperation())
+        .Case<memref::AllocaOp>([](auto) { return true; })
+        .Case<memref::StoreOp, affine::AffineStoreOp>([this](auto storeOp) {
+          UniformityLattice *lattice =
+              getLatticeElement(storeOp.getValueToStore());
+          return !lattice->getValue().isUninitialized();
         })
-        .Case<MemoryEffects::Write>([](auto) {
-          // A write operation store a value to a memory location. Anything
-          // to do here ?
-          assert(false && "TODO");
+        .Default([](auto) {
+          llvm_unreachable("Unhandled operation kind");
+          return false;
```

Comment:
```suggestion
```

Reply:
I could not remove the return stmt. Got an error from g++:
```
/nfs/site/home/etiotto/projects/intel-llvm/polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:292:17:   required from here
/nfs/site/home/etiotto/projects/intel-llvm/llvm/include/llvm/ADT/TypeSwitch.h:131:21: error: void value not ignored as it ought to be
  131 |     return defaultFn(this->value);
      |            ~~~~~~~~~^~~~~~~~~~~~~
```